spec/controllers/reports_controller_spec.rb

```diff
@@ -0,0 +1,334 @@
+require "rails_helper"
+include SessionsHelper
+
+RSpec.describe ReportsController, type: :controller do
+
+  let!(:user) {FactoryBot.create :user, id: 1}
+  let!(:user_manager) {FactoryBot.create :user, id: 2, role: :normal}
+  let!(:user_admin) {FactoryBot.create :user, id: 3, role: :admin}
+  let!(:user_singe) {FactoryBot.create :user, id: 4, role: :normal}
```

Comment:
sao lại dùng let! thế này
khác nhau giữa let và let! là gì e?

Reply:
Sự khác nhau giữa **let** và **let!**
**let** chỉ chạy khi và chỉ khi nào được gọi đến, chỉ có giá trị trong từng example. (trong Rspec thì được đóng gói trong khối it) và nếu như gọi đến một biến let nhiều lần trong cùng một example thì nó sẽ chỉ khai báo khối let một lần duy nhất (ở lần gọi đầu tiên). Đây là khái niệm lazy evaluation. 
**let!** được gọi trong một khối before ngầm, do đó kết quả đã được tính toán và cache lại trước khi khối it được thực thi.
Ở trường hợp này em dùng let! vì data được dùng chung cho tất cả các example.